docs/proposals/network-policy.md

```diff
@@ -0,0 +1,316 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# NetworkPolicy
+
+## Abstract
+
+A proposal for implementing a new resource - NetworkPolicy - which
+will enable definition of ingress policies for selections of pods.
+
+The design for this proposal has been created by, and discussed
+extensively within the Kubernetes networking SIG.  It has been implemented
+and tested using Kubernetes API extensions by various networking solutions already.
+
+In this design, users can create various NetworkPolicy objects which select groups of pods and
+define how those pods should be allowed to communicate with each other.  The
+implementation of that policy at the network layer is left up to the
+chosen networking solution.
+
+> Note that this proposal does not yet include egress / cidr-based policy, which is still actively undergoing discussion in the SIG. These are expected to augment this proposal in a backwards compatible way.
+
+## Implementation
+
+The implmentation in Kubernetes consists of:
+- A v1beta1 NetworkPolicy API object
+- A structure on the `Namespace` object to control policy, to be developed as an annotation for now.
+
+### Namespace changes
+
+The following objects will be defined on a Namespace Spec.
+>NOTE: In v1beta1 these objects will be implemented as an annotation.
+
+```go
+type IngressIsolationPolicy string
+
+const (
+	// Deny all ingress traffic to pods in this namespace.
+	DefaultDeny IngressIsolationPolicy = "DefaultDeny"
+) 
+
+// Standard NamespaceSpec object, modified to include a new
+// NamespaceNetworkPolicy field.
+type NamespaceSpec struct {
+	// This is a pointer so that it can be left undefined.
+	NetworkPolicy *NamespaceNetworkPolicy `json:"networkPolicy,omitempty"`
```

Comment:
I'm fine with KISS - we're starting to see substantial use of namespaces
and this is the inevitable fallout, nothing here precludes a different
solution.

On Fri, May 6, 2016 at 1:51 AM, Tim Hockin <notifications@github.com> wrote:

> In docs/proposals/network-policy.md
> <https://github.com/kubernetes/kubernetes/pull/24154#discussion_r62290997>
> :
>
> > +The following objects will be defined on a Namespace Spec.
> > +>NOTE: In v1beta1 these objects will be implemented as an annotation.
> > +
> > +```go
> > +type IngressIsolationPolicy string
> > +
> > +const (
> > +	// Deny all ingress traffic to pods in this namespace.
> > +	DefaultDeny IngressIsolationPolicy = "DefaultDeny"
> > +)
> > +
> > +// Standard NamespaceSpec object, modified to include a new
> > +// NamespaceNetworkPolicy field.
> > +type NamespaceSpec struct {
> > +	// This is a pointer so that it can be left undefined.
> > +	NetworkPolicy *NamespaceNetworkPolicy `json:"networkPolicy,omitempty"`
>
> I have chewed on the idea of label-selecting namespaces for tenant IDs
> (applied by admission, not by users, obv.) and using that to apply a
> higher-level of network policy that bounds the user-defined. It's not clear
> if that means a new global object Kind or an administrative namespace that
> uses the existing NetworkPolicy Kind (and the implications of that on this
> proposal).
>
> TL;DR I think it's doable, but we wanted, NEEDED to KISS here.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/24154/files/bc3e88127e5c75e29781f70d5a1b8697cbc0702b#r62290997>
>